### PR TITLE
check for invalid language and data type QIDs

### PR DESCRIPTION
**Pull request by DeleMike**
### Contributor checklist

- [x] This pull request is on a [separate branch](https://docs.github.com/en/get-started/quickstart/github-flow) and not the main branch

---

### Description

I am showing a draft of how we want to verify queries. That is, language QIDs and data type QIDs. 

### Related issue

- #339

### CODE REVIEW

**Comment by DeleMike** [comment]
### Some observations

- The issue with language QID verification is we don't have an updated source of truth.

- The issue with data type QID verification is the queries are written in different manners. 

It will be hard to extract the QIDs. For example see `src/scribe_data/language_data_extraction/English/nouns/query_nouns.sparql`:
```sparql
# tool: scribe-data
# All English (Q1860) nouns and their plural.
# Enter this query at https://query.wikidata.org/.

SELECT DISTINCT
  (REPLACE(STR(?lexeme), "http://www.wikidata.org/entity/", "") AS ?lexemeID)
  ?singular
  ?plural

WHERE {
  VALUES ?nounTypes {wd:Q1084 wd:Q147276} # Nouns and proper nouns

  ?lexeme dct:language wd:Q1860 ;
    wikibase:lexicalCategory ?nounTypes ;
    wikibase:lemma ?singular .

  # MARK: Plural

  OPTIONAL {
    ?lexeme ontolex:lexicalForm ?pluralForm .
    ?pluralForm ontolex:representation ?plural ;
      wikibase:grammaticalFeature wd:Q146786 ;
  } .
}
```
and see verbs
```sparql
# tool: scribe-data
# All Nigerian Pidgin (Q33655) verbs.
# Enter this query at https://query.wikidata.org/.

SELECT
  (REPLACE(STR(?lexeme), "http://www.wikidata.org/entity/", "") AS ?lexemeID)
  ?verb

WHERE {
  ?lexeme dct:language wd:Q33655 ;
    wikibase:lexicalCategory wd:Q24905 ;
    wikibase:lemma ?verb .
}
```

the `lexicalCategory wd:Q24905` is different for verbs and nouns.

File: src/scribe_data/check/check_query_identifiers.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        print("Queries with incorrect data type QIDs are:")
+        for file in incorrect_data_types:
+            print(f"- {file}")
+
+
+def is_valid_language(query_file, lang_qid):
+    lang_directory_name = query_file.parent.parent.name.lower()
+    languages = language_metadata.get(
+        "languages"
+    )  # might not work since language_metadata file is not fully updated
```
Comment (DeleMike) [inline src/scribe_data/check/check_query_identifiers.py:58]: For language verification, the stumbling block might be the `language_metadata.json`. If it is updated properly then we won't have issues. It depends of #293

**Comment by DeleMike** [comment]
> ### Some observations
> * The issue with language QID verification is we don't have an updated source of truth.
> * The issue with data type QID verification is the queries are written in different manners.
> 
> It will be hard to extract the QIDs. For example see `src/scribe_data/language_data_extraction/English/nouns/query_nouns.sparql`:
> 
> ```sparql
> # tool: scribe-data
> # All English (Q1860) nouns and their plural.
> # Enter this query at https://query.wikidata.org/.
> 
> SELECT DISTINCT
>   (REPLACE(STR(?lexeme), "http://www.wikidata.org/entity/", "") AS ?lexemeID)
>   ?singular
>   ?plural
> 
> WHERE {
>   VALUES ?nounTypes {wd:Q1084 wd:Q147276} # Nouns and proper nouns
> 
>   ?lexeme dct:language wd:Q1860 ;
>     wikibase:lexicalCategory ?nounTypes ;
>     wikibase:lemma ?singular .
> 
>   # MARK: Plural
> 
>   OPTIONAL {
>     ?lexeme ontolex:lexicalForm ?pluralForm .
>     ?pluralForm ontolex:representation ?plural ;
>       wikibase:grammaticalFeature wd:Q146786 ;
>   } .
> }
> ```
> 
> and see verbs
> 
> ```sparql
> # tool: scribe-data
> # All Nigerian Pidgin (Q33655) verbs.
> # Enter this query at https://query.wikidata.org/.
> 
> SELECT
>   (REPLACE(STR(?lexeme), "http://www.wikidata.org/entity/", "") AS ?lexemeID)
>   ?verb
> 
> WHERE {
>   ?lexeme dct:language wd:Q33655 ;
>     wikibase:lexicalCategory wd:Q24905 ;
>     wikibase:lemma ?verb .
> }
> ```
> 
> the `lexicalCategory wd:Q24905` is different for verbs and nouns.

@andrewtavis what do you think of this?

The language issue is kind of sorted. Once we have an updated language_metadata file then this code should work as expected.

However, verifying data_types is difficult because we have different ways the SPARQL queries were constructed.

PS: Please ignore some of the print statements, I was using them to debug. This is still a draft PR.

File: src/scribe_data/check/check_query_identifiers.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            if match:
+                return match.group(0).replace("wd:", "")
+    except Exception as e:
+        print(f"Error reading {file_path}: {e}")
+    return None
+
+
+def check_queries():
+    language_pattern = r"\?lexeme dct:language wd:Q\d+"
+    data_type_pattern = r"wikibase:lexicalCategory wd:Q\d+"
```
[the remark this answers is not in the document]
Comment (andrewtavis) [reply src/scribe_data/check/check_query_identifiers.py:28]: We can certainly split the current preposition and postposition queries, @DeleMike :) Do you want to send along a PR for that?

As far as nouns and proper nouns, how are we feeling on this? Split them too? CC @catreedle :)

**Comment by andrewtavis** [comment]
As stated above, if we want to make the switch to one data type per query then I'm fine with that :)

**Comment by andrewtavis** [comment]
There's potentially more value in making sure that we can maintain the query base than combining nouns and proper nouns, plus some people might only want one 🤔

**Comment by catreedle** [comment]
How about we keep a dictionary of the patterns for data type?

```py
data_type_patterns = {
    "nouns": r"\?nounTypes \{wd:(Q\d+)\}", 
    "default": r"wikibase:lexicalCategory wd:Q\d+"
}
```

We can retrieve the pattern based on the directory name
```py
directory_name = query_file.parent.name
data_type_pattern = data_type_patterns.get(directory_name, data_type_patterns["default"])  # Use default if no match
```

This allows us to manage different patterns for various data types. If a directory name matches a key in the dictionary, we get its specific pattern, otherwise we use a default pattern for other types.
What do you think?

**Comment by andrewtavis** [comment]
I think that postpositions and prepositions should be split regardless, but this could certainly work 😊

File: src/scribe_data/check/check_query_identifiers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (DeleMike) [reply src/scribe_data/check/check_query_identifiers.py:28]: Yes, I agree. we need to split them

**Comment by DeleMike** [comment]
@catreedle your idea would work but the real issue might still be there.  So as @andrewtavis says, we should convert the queries to have one query per type. It would make the validation process easier. This pattern will now stand for all queries if we split all the queries:
```sparql 
data_type_pattern = r"wikibase:lexicalCategory wd:Q\d+"
```
Hence, in this style of code:
```sparql
SELECT
  (REPLACE(STR(?lexeme), "http://www.wikidata.org/entity/", "") AS ?lexemeID)
  ?preposition
  ?case

WHERE {
  # Prepositions and postpositions.
  VALUES ?prePostPositions { wd:Q4833830 wd:Q161873 }

  ?lexeme dct:language wd:Q9610 ;
    wikibase:lexicalCategory ?prePostPositions ;
    wikibase:lemma ?preposition .

  # MARK: Corresponding Case

  OPTIONAL {
    ?lexeme wdt:P5713 ?caseForm .
  } .

  SERVICE wikibase:label {
    bd:serviceParam wikibase:language "[AUTO_LANGUAGE]".
    ?caseForm rdfs:label ?case .
  }
}
```

we should eliminate all `VALUES ?...` and then have:
`query_preposition.sparql`:
```sparql
SELECT DISTINCT
  ?lexeme
  (REPLACE(STR(?lexeme), "http://www.wikidata.org/entity/", "") AS ?lexemeID)
  ?preposition

WHERE {
  ?lexeme dct:language wd:Q9610 ;
    wikibase:lexicalCategory wd:Q4833830 ;
    wikibase:lemma ?preposition .
    FILTER(lang(?preposition) = "hi")
}
```

`query_postposition.sparql`:
```sparql
SELECT DISTINCT
  ?lexeme
  (REPLACE(STR(?lexeme), "http://www.wikidata.org/entity/", "") AS ?lexemeID)
  ?postposition

WHERE {
  ?lexeme dct:language wd:Q9610 ;
    wikibase:lexicalCategory wd:Q161873 ;
    wikibase:lemma ?postposition .
    FILTER(lang(?postposition) = "hi")
}
```


However, this might lead to a large PR because there would be many shake-ups.

@andrewtavis is there a need for me to create an issue? And what do you think about this?

**Comment by andrewtavis** [comment]
I'm totally fine with this if this means that it's a bit easier for us to make sure that everything is working properly. Plus again there's a chance that someone just wants proper nouns. I can even think of an example right now where someone wants a list of names in a language, and then they'd be able to use Scribe-Data to easily get all names of a given language :)

Should this be another issue?

**Comment by DeleMike** [comment]
Thanks @andrewtavis!

Yes, this should be tracked as a separate issue. It’s a significant refactor that involves modifying the structure of multiple query files and creating a dedicated issue would help us manage the process effectively.

I’ll create an issue, and then get started on a PR asap so this **current PR** can make progress.

**Comment by andrewtavis** [comment]
Sounds good, Thanks @DeleMike!

**Comment by andrewtavis** [comment]
Do you want to clean this PR up as well, @DeleMike, and then I can do a final review?

**Comment by DeleMike** [comment]
> Do you want to clean this PR up as well, @DeleMike, and then I can do a final review?

I will clean it up, but we can't say for sure if everything will be 100% good to go with the checks, even after resolving #380.

Would you like me to adjust the current PR with the new plan in mind, where the pattern will be updated to:

```python
data_type_pattern = r"wikibase:lexicalCategory wd:Q\d+"
```

This will ensure that our refactor is aligned with the upcoming changes.

**Comment by DeleMike** [comment]
I will focus on cleaning up `src/scribe_data/check/check_query_identifiers.py` to have the right pattern and flow to achieve these checks. 
Then, I will jump straight into configuring the queries.
Once that is set up, I will run the checks again to verify the queries.

You should be expecting a clean PR in less than 30 mins :)

**Comment by andrewtavis** [comment]
Thanks much, @DeleMike, and yes I think that it makes sense to refactor based on the changes in #380 going through 😊 We can merge the work for the other one before this one :)

**Comment by DeleMike** [comment]
currently facing merge conflicts...

**Comment by DeleMike** [comment]
@andrewtavis I resolved the merge conflicts. Please help review to know if this is acceptable for now🤔

File: src/scribe_data/check/check_query_identifiers.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+def check_queries():
+    """
+    Validates SPARQL queries in the specified directory to check for correct language
+    and data type QIDs.
+
+    This function scans all SPARQL query files in the LANGUAGE_DATA_EXTRACTION_DIR
+    and prints out any files with incorrect QIDs for both languages and data types.
+    """
+    language_pattern = r"\?lexeme dct:language wd:Q\d+"
+    data_type_pattern = r"wikibase:lexicalCategory\s+wd:Q\d+"
```
Comment (DeleMike) [inline src/scribe_data/check/check_query_identifiers.py:47]: upgraded the regex expression

Comment (DeleMike) [reply src/scribe_data/check/check_query_identifiers.py:47]: and docstrings. Those were the major work in the commit.

are the docstrings in the right format? @andrewtavis

Comment (andrewtavis) [reply src/scribe_data/check/check_query_identifiers.py:47]: Yes they're looking good now, @DeleMike :) I'd just indent parameter and return entries, but really good 😊

**Review by andrewtavis — APPROVED** [review]
This is great, @DeleMike! Looking forward to seeing it in action in a workflow 😊

**Pull request merged**